pkg/imageupdater/imageupdater.go

```diff
@@ -0,0 +1,187 @@
+package imageupdater
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/nais/knorten/pkg/chart"
+	"github.com/nais/knorten/pkg/database"
+	"github.com/nais/knorten/pkg/database/crypto"
+	"github.com/nais/knorten/pkg/database/gensql"
+	"github.com/nais/knorten/pkg/k8s"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	profileListHelmKey = "singleuser.profileList"
+)
+
+type ImageUpdater struct {
+	repo          *database.Repo
+	jupyterClient *chart.JupyterhubClient
+	log           *logrus.Entry
+}
+
+type garImage struct {
+	Name string `json:"package"`
+	Tag  string `json:"tags"`
+}
+
+type profile struct {
+	Name                string `json:"display_name"`
+	Description         string `json:"description"`
+	KubespawnerOverride struct {
+		Image string `json:"image"`
+	} `json:"kubespawner_override"`
+}
+
+func New(repo *database.Repo, k8sClient *k8s.Client, cryptClient *crypto.EncrypterDecrypter, jupyterChartVersion string, log *logrus.Entry) *ImageUpdater {
+	jupyterClient := chart.NewJupyterhubClient(repo, k8sClient, cryptClient, jupyterChartVersion, log.WithField("subsystem", "jupyterClient"))
+	return &ImageUpdater{
+		repo:          repo,
+		jupyterClient: &jupyterClient,
+		log:           log,
+	}
+}
+
+func (d *ImageUpdater) Run(frequency time.Duration) {
+	ctx := context.Background()
+
+	ticker := time.NewTicker(frequency)
+	defer ticker.Stop()
+	for {
+		d.run(ctx)
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+		}
+	}
+}
+
+func (d *ImageUpdater) run(ctx context.Context) {
+	profilesDB, err := d.repo.GlobalValueGet(ctx, gensql.ChartTypeJupyterhub, profileListHelmKey)
+	if err != nil {
+		d.log.WithError(err).Error("getting jupyterhub singleuser profiles from database")
+		return
+	}
+
+	profiles := []*profile{}
+	if err := json.Unmarshal([]byte(profilesDB.Value), &profiles); err != nil {
+		d.log.WithError(err).Error("unmarshalling profiles")
+		return
+	}
+
+	profilesStatus := []bool{}
+	for _, p := range profiles {
+		updated, err := updateIfNeeded(p)
+		if err != nil {
+			d.log.WithError(err).Error("checking image up to date with GAR")
+		}
+		profilesStatus = append(profilesStatus, updated)
```

Comment:
Hvorfor ikke bare kalle på det som skjer i if-en på linje 91 med en gang her? Spiller det noen rolle hvor mange som er true?

Reply:
Mulig jeg misforstår og er kanskje dårlig løsning, men `updateIfNeeded` på linje 84 vil oppdatere imaget hvis det er behov i den lokale listen over profiler. Gjorde det slik for å slippe å måtte skrive global valuen til vår database flere ganger (da denne inneholder alle profilene). Det er ikke sikkert at alle profilene (image med forskjellig python versjon) skal oppdateres. Når workflowen trigger i [navikt/knada-images](https://github.com/navikt/knada-images) på bakgrunn av endring i jupyter/base-notebook så kan det hende at det kun gjelder for f.eks. python-3.10 versjonen og da bygges ikke de resterende.